src/ratchet.cpp

```diff
@@ -584,15 +590,21 @@ std::size_t olm::Ratchet::decrypt(
                 }
             }
         }
+        /* No key found for this message. */
+        last_error = OlmErrorCode::OLM_ALREADY_DECRYPTED_OR_KEYS_SKIPPED;
     } else {
         result = verify_mac_and_decrypt_for_existing_chain(
             *this, chain->chain_key,
-            reader, plaintext, max_plaintext_length
+            reader, plaintext, max_plaintext_length,
+            last_error
         );
     }
 
     if (result == std::size_t(-1)) {
-        last_error = OlmErrorCode::OLM_BAD_MESSAGE_MAC;
+        /* There was an error but specific error code wasn't set.  */
+        if(last_error == OlmErrorCode::OLM_SUCCESS) {
```

Comment:
1. Can you add a space after `if`?
2. I'm not sure about this condition. Is there a possibility that `last_error` was not set in this invocation, and so we end up checking the `last_error` from the previous invocation here? If so, it seems like we might accidentally not set `last_error` here

Reply:
1. Done
2. I don't think it's possible, each time we "consume" `last_error` from [olm::Ratchet](https://github.com/CommE2E/olm/blob/9d51b9c0d478d1adc1efff1f09298c3d7e2702db/include/olm/ratchet.hh#L93) which is in those places:
 - https://github.com/CommE2E/olm/blob/9d51b9c0d478d1adc1efff1f09298c3d7e2702db/src/session.cpp#L354
 - https://github.com/CommE2E/olm/blob/9d51b9c0d478d1adc1efff1f09298c3d7e2702db/src/session.cpp#L388
 - https://github.com/CommE2E/olm/blob/9d51b9c0d478d1adc1efff1f09298c3d7e2702db/src/session.cpp#L423
 We set it back to `OlmErrorCode::OLM_SUCCESS`. We also do it on [olm::Ratchet constructor](https://github.com/CommE2E/olm/blob/31fd0e0730c92beb7ad7655dafba1a726eeade6a/src/ratchet.cpp#L192) so my interpretation is that on each invocation we should start with `last_error == OlmErrorCode::OLM_SUCCESS` and then when `result` is `-1` we have a specific error or `OLM_BAD_MESSAGE_MAC`.